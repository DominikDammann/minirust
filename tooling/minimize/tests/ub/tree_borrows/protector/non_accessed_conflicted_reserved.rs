//@ compile-flags: --minimize-tree-borrows

// Check that a foreign read makes future child writes to this mutable reference UB, even though the reference is non-accessed.

fn foo(x: *mut u8, y: &mut u8) {
    // (x, [R, R]) -> (y, P[[R, R]])
    let yraw = y as *mut u8;
    unsafe {
        assert!(*x.add(1) == 57); // (x, [R, R]) -> (y, P[[R, RC]])
        assert!(*yraw.add(1) == 57); // y[1] becomes Accessed (Actively protected)
        *yraw.add(1) = 42; // UB! Child Write to Actively Protected Conflicted Reserved.
    } 
}

fn main() {
    let mut data = [42u8, 57]; 
    let x = &mut data[0] as *mut u8; // (x, [R, R]) 
    let y = unsafe { &mut *x }; // (x, [R, R]) -> (y, [R, R])
    foo(x, y);
}
