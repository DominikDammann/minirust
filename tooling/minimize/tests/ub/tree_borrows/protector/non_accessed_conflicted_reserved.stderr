fatal error: UB: Tree Borrows: writing to the local of a protected pointer with Conflicted Reserved permission
